Add a picture uploader to picture-card-editor

karwosts opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

Add a picture-upload form to the picture-card editor. I think this makes working with images much more accessible, otherwise it can be a source of a lot of confusion how to add pictures to the server and figure out the path. 

Maybe this should be wrapped up in some kind of image-selector element that can be reused? 

That would allow for either entering an external URL, uploading, or maybe opening a file picker and allowing to select from the list of current uploaded images either in /image/ or /www/?

![image-picker](https://github.com/home-assistant/frontend/assets/32912880/18035436-fa62-4bb8-b54f-4262b9b98830)

## Type of change
<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
https://github.com/home-assistant/frontend/discussions/18669
- Link to documentation pull request:

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - Introduced an image selector component allowing users to upload or specify an image via URL.
  - Added new image-related options in various card editors for enhanced customization.

- **Style**
  - Updated button styles to better display warning buttons with a distinct color.

- **Translations**
  - Added new English translations for image selection, upload, and URL input options.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

karwosts committed:
Add a picture uploader to picture-card-editor

bramkragten commented:
Oh I do like this!

> Maybe this should be wrapped up in some kind of image-selector element that can be reused?

Yes, that would be a good idea, similar to the `file selector`

> That would allow for either entering an external URL, uploading, or maybe opening a file picker and allowing to select from the list of current uploaded images either in /image/ or /www/?

That also sounds good, but should be options on the selector then probably? The last option would need backend support first. (We do need something similar for uploaded pictures though probably, as there is no way to manage or reuse already uploaded images)

karwosts commented:
If we had image-selector, what do you think of how the current UI looks? 

OK to just have a `<ha-picture-upload>` stacked on top of a textbox for the path. Or is that take up too much space? Maybe have to click a button to show the uploader?

Image browser/picker could come later if/when that ever gets implemented.

matthiasdebaat commented:
I like this one too! I'm seeing 3 options to select an image, correct me if I'm wrong:
- Select a local file using image URL
- Upload image from device, that converts to a local image URL
- Upload image from web, that converts to a local image URL

In my opinion, the input-select with image URL isn't user friendly and pretty advanced. You have to know the path to a file, that requires access to your system. Can we drop this one? That means there should be an option to select an image from the system, like we're doing with the Media panel.

karwosts commented:
> I like this one too! I'm seeing 3 options to select an image, correct me if I'm wrong:
> - Select a local file using image URL

The URL input can also be remote URL on the web, it doesn't have to be local.

bramkragten commented:
I think I would hide the input by default, but allow the user to switch to it with a button

matthiasdebaat commented:
> > I like this one too! I'm seeing 3 options to select an image, correct me if I'm wrong:
> > 
> > * Select a local file using image URL
> > * Upload image from device, that converts to a local image URL
> > * Upload image from web, that converts to a local image URL
> 
> The URL input can also be remote URL on the web, it doesn't have to be local.

That is the third option in my mind. So it would be:
*  Upload image from web, that converts web URL to a local image URL

matthiasdebaat commented:
> I think I would hide the input by default, but allow the user to switch to it with a button

I agree. There should be a way to select between selecting a local (already uploaded) file, upload new image or upload image from the web. This last option only have an input. The other doesn't need one in my mind.

karwosts commented:
> I think I would hide the input by default

Can you clarify, what is "the input" you're referring to hide, the text selector, or the picture-upload?

I was kind of thinking to have the button toggle the larger upload element, though I'm wondering if Matthias is suggesting the opposite; to hide the URL. I'm not sure if that makes sense to me yet. 

This is kind of what I was going to go for:

![picture-card-2](https://github.com/home-assistant/frontend/assets/32912880/36a04563-bfe3-427f-8591-22893a11ef6a)



Matthias, you seem to also be suggesting that remote URLs should be replaced with "upload to a local image from web", but I think people might not necessarily want to upload remote URLs. The behavior now is just to show the live image from the remote domain. Maybe that is desired, in case the remote image is dynamic, or user just don't want to store it on his server?

karwosts committed:
add imageSelector

karwosts committed:
lint

bramkragten commented:
@matthiasdebaat can you check above questions?

matthiasdebaat commented:
Thanks for your feedback. Based on this I've got something like this in my mind. More details can be found in the [Figma design](https://www.figma.com/file/DvtgSpuwNKQvSIWTvBxTn7/Picture-card-upload-image?type=design&node-id=0%3A1&mode=design&t=tldtNKmfZsYN3gq8-1).

![CleanShot 2023-12-07 at 10 59 49@2x](https://github.com/home-assistant/frontend/assets/20203284/d800c2ea-dec6-467b-b9aa-e13264d24c96)

My suggestions are:
- Let the user to choose between upload or URL. When the user chooses upload, the URL input is hidden and vice versa.
- Add a delete button when an image is uploaded.
- URL is as we currently have it.

karwosts committed:
Add delete button to picture-upload

karwosts committed:
updates from feedback

karwosts commented:
Updated from feedback.

I've for now dropped the `crop` feature from this as I don't really like how it works. When crop is enabled you are forced to go through the crop dialog, and there's no easy way to just use the entire image unaltered (you would have to try to drag the crop bubbles to exactly the corner of your image, and probably get it off by 1 pixel), so that seems potentially frustrating. 

A better flow might be to have first the entire picture uploaded without crop, and then have a `CROP` button on the line below the picture to run it through the crop flow? Maybe that could be a different change. 


![image-selector-3](https://github.com/home-assistant/frontend/assets/32912880/15e4e32d-2cbb-41a0-a252-68b98869f6a0)

karwosts committed:
fix lint

bramkragten commented on src/components/ha-picture-upload.ts:
```diff
@@ -60,13 +68,23 @@ export class HaPictureUpload extends LitElement {
+            )}
+          >
+          </ha-button>
+          ${this.canDelete
```
I don't think it is clear to a user what the delete button does, it looks like it does the same as the change picture button, but it deletes it from the server...

We should have a picker of current uplloaded files sometime 😓

karwosts replied:
I would offer to remove the delete button, but it was just added on Matthias's request 😕 

The other weird thing about delete is that it's in a cancellable dialog. If you delete the picture, but then cancel the change, the picture becomes a broken link to a no longer existing image.

I assumed that more picture management functionality would maybe come someday in the future, but I felt maybe this was ok to go ahead without that, as that felt like might be a pretty big project with new APIs.

bramkragten replied:
@matthiasdebaat

matthiasdebaat replied:
For the sake of not blocking this PR, I'm okay if the delete button is removed.

> We should have a picker of current uplloaded files sometime 😓

I agree

karwosts replied:
> For the sake of not blocking this PR, I'm okay if the delete button is removed.

It's gone.

karwosts committed:
Update en.json

karwosts committed:
Update selector.ts

karwosts committed:
remove delete

bramkragten commented:
I would still really like to combine the image entity/camera entity/upload image/image url options for improved clarity that they can not be used together. But this is already a nice addition that we should merge. Thanks @karwosts!
